lib/elixir/lib/module/locals_tracker.ex

```diff
@@ -104,13 +105,31 @@ defmodule Module.LocalsTracker do
   def collect_undefined_locals({set, bag}, all_defined) do
     undefined =
       for {pair, _, _, _} <- all_defined,
-          {local, line} <- out_neighbours(bag, {:local, pair}),
+          {local, line, _} <- out_neighbours(bag, {:local, pair}),
           not :ets.member(set, {:def, local}),
           do: {build_meta(line), local}
 
     :lists.usort(undefined)
   end
 
+  @doc """
+  Collects incorrect dispatches based on local calls and existing
+  definitions.
+
+  An incorrect dispatch occurs when a macro is invoked before its
+  definition.
+  """
+  def collect_incorrect_dispatches({set, bag}, all_defined) do
```

Comment:
Unfortunately this requires you to traverse all definitions and all calls.

I wonder if instead we should have a new entry in the table like this:

    {:macro_call, caller, callee}

And  then we just need to traverse the calls. Thoughts?

Reply:
This function detects "function calls" that should have been dispatched as "macro calls". I think the structure should be then `{:function_call, caller, callee}`.

Regarding performance, I thought about merging all the `collect_*` functions in the locals tracker so we only traverse definitions and calls once, but I was thinking on doing that in a different PR 🙂 

Thoughts?

Reply:
The other `collect` functions end-up looking up for different data, so I wouldn't worry about merging them, but this could definitely be merged with `unused_locals`. They are even part of the same problem. So what if we merge those two already and have it return something like `{pair, line, reason}` as you proposed? Then I don't think we need to do the changes I mentioned above.  :+1:

Reply:
> but this could definitely be merged with `unused_locals`

This was prevously raising an “undefined function error”, so I was thinking on merging it with `collect_undefined_locals`.

Reply:
Sorry, I meant indeed to merge with `collect_undefined_locals`, that's the closest one. :)